cylc/flow/config.py

```diff
@@ -140,23 +146,23 @@ class WorkflowConfig:
 
     def __init__(
         self,
-        workflow,
-        fpath,
-        options=None,
-        template_vars=None,
-        is_reload=False,
-        output_fname=None,
-        xtrigger_mgr=None,
-        mem_log_func=None,
-        run_dir=None,
-        log_dir=None,
-        work_dir=None,
-        share_dir=None,
-    ):
+        workflow: str,
+        fpath: str,
+        options: Optional['Values'] = None,
+        template_vars: Optional[Mapping[str, Any]] = None,
+        is_reload: bool = False,
+        output_fname: Optional[str] = None,
+        xtrigger_mgr: Optional[XtriggerManager] = None,
+        mem_log_func: Optional[Callable] = None,
+        run_dir: Optional[str] = None,
+        log_dir: Optional[str] = None,
+        work_dir: Optional[str] = None,
+        share_dir: Optional[str] = None
+    ) -> None:
 
         self.mem_log = mem_log_func
-        if mem_log_func is None:
-            self.mem_log = lambda *a: False
+        if self.mem_log is None:
+            self.mem_log = lambda *a, **k: False
```

Comment:
Not sure I understand why Mypy needs this, but it looks weird given that we never call this with keyword args.

It seems to be possible to annotate lambdas properly. This passes for me locally:
```diff
diff --git a/cylc/flow/config.py b/cylc/flow/config.py
index 9691a5329..63c44d29e 100644
--- a/cylc/flow/config.py
+++ b/cylc/flow/config.py
@@ -153,16 +153,17 @@ class WorkflowConfig:
         is_reload: bool = False,
         output_fname: Optional[str] = None,
         xtrigger_mgr: Optional[XtriggerManager] = None,
-        mem_log_func: Optional[Callable] = None,
+        mem_log_func: Optional[Callable[[str], bool]] = None,
         run_dir: Optional[str] = None,
         log_dir: Optional[str] = None,
         work_dir: Optional[str] = None,
         share_dir: Optional[str] = None
     ) -> None:
 
-        self.mem_log = mem_log_func
-        if self.mem_log is None:
-            self.mem_log = lambda *a, **k: False
+        if mem_log_func is None:
+            self.mem_log: Callable[[str], bool] = lambda *a: False
+        else:
+            self.mem_log = mem_log_func
         self.mem_log("config.py:config.py: start init config")
         self.workflow = workflow  # workflow name
         self.fpath = fpath  # workflow definition
```